Library/Homebrew/utils/io_selector.rb

```diff
@@ -0,0 +1,62 @@
+require "delegate"
+require "English"
+
+require "extend/io"
+
+module Utils
+  class IOSelector < DelegateClass(Hash)
+    attr_reader :separator
+
+    alias all_streams keys
+    alias all_tags values
+    alias tag_of fetch
+
+    def self.each_line_from(streams = {}, separator = $RS, &block)
+      new(streams, separator).each_line_nonblock(&block)
+    end
+
+    def initialize(streams = {}, separator = $RS)
+      super(streams.invert.compare_by_identity)
+      @separator = separator
+    end
+
+    def each_line_nonblock
```

Comment:
Is `each_line_nonblock` the right name for this method? It calls `readline_nonblock`, which is nonblocking, but `each_readable_stream_until_eof` is calling `IO::select`, which may block if there is no input on any stream, so this method may block. Maybe `each_line_each_stream` or `each_line_from_my_streams` or `each_line_dont_deadlock`?